File: day1/1_hall/hall_as.java
import java.io.*;
import java.util.*;

public class hall_as {
    public static void main(String[] s) throws IOException {
        new hall_as().run();
    }

    void run() throws IOException {
        Scanner in = new Scanner(new File("hall.in"));
        PrintWriter out = new PrintWriter(new File("hall.out"));

        int a = in.nextInt();
        int b = in.nextInt();
        int c = in.nextInt();
        int d = in.nextInt();

        long ans = 0;
        for (int x = 1; x <= d / 2 && x * x <= b; x++) {
            int miny = x;
            if ((c + 1) / 2 - x > miny) {
                miny = (c + 1) / 2 - x;
            }
            if ((a + x - 1) / x > miny) {
                miny = (a + x - 1) / x;
            }
            int maxy = d / 2 - x;
            if (b / x < maxy) {
                maxy = b / x;
            }
            if (maxy >= miny) {
                ans += maxy - miny + 1;
            }
        }
        out.println(ans);

        in.close();
        out.close();
    }
}

File: day1/1_hall/hall_as.pas
var
    a, b, c, d: longint;
    x, miny, maxy: longint;
    ans: int64;
begin
    assign(input, 'hall.in');
    reset(input);
    assign(output, 'hall.out');
    rewrite(output);

    read(a, b, c, d);
    ans := 0;

    for x := 1 to d div 2 do begin
        if x * x > b then
            break;
        miny := x;
        if (c + 1) div 2 - x > miny then
            miny := (c + 1) div 2 - x;
        if (a + x - 1) div x > miny then
            miny := (a + x - 1) div x;
        maxy := d div 2 - x;
        if b div x < maxy then
            maxy := b div x;
        if maxy >= miny then
            ans := ans + (maxy - miny + 1);
    end;
    writeln(ans);

    close(input);
    close(output);
end.
